File: py/structure/tree/search_tree/BinarySearchTree.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys

class BinarySearchTree:
    def __init__(self):
        self.__root = None

    def isEmpty(self) -> bool:
        return not self.__root

    def add(self, key: int, value: int) -> None:
        if not self.__root:
            self.__root = self.__TreeNode(key, value)
            return

        parent, current = None, self.__root
        while current:
            if key == current.key:
                current.value = value
                return

            parent = current
            if key < current.key:
                current = current.left
            else:
                current = current.right
        self.__addNode(parent, key, value)

    def erase(self, key: int) -> None:
        if not self.__root:
            return

        if key == self.__root.key and (not self.__root.left or not self.__root.right):
            self.__root = self.__root.left if self.__root.left else self.__root.right
            return

        parent, current = None, self.__root
        while current:
            if key == current.key:
                break

            parent = current
            if key < current.key:
                current = current.left
            else:
                current = current.right

        if not current:
            return

        if not current.left or not current.right:
            self.__eraseNode(parent, current, current.left if current.left else current.right)
            return

        child, parent = current.right, current
        while child.left:
            parent, child = child, child.left
        current.key, current.value = child.key, child.value
        self.__eraseNode(parent, child, child.right)

    def set(self, key: int, value: int) -> None:
        current = self.__root
        while current:
            if key == current.key:
                current.value = value
                return

            if key < current.key:
                current = current.left
            else:
                current = current.right

    def get(self, key: int) -> int:
        current = self.__root
        while current:
            if key == current.key:
                return current.value

            if key < current.key:
                current = current.left
            else:
                current = current.right
        return -sys.maxsize

    def clear(self) -> None:
        self.__root = None

    def __addNode(self, parent: '__TreeNode', key: int, value: int) -> None:
        if key < parent.key:
            parent.left = self.__TreeNode(key, value)
        else:
            parent.right = self.__TreeNode(key, value)

    def __eraseNode(self, parent: '__TreeNode', current: '__TreeNode', child: '__TreeNode') -> None:
        if current == parent.left:
            parent.left = child
        else:
            parent.right = child

    class __TreeNode:
        def __init__(self, key: int, value: int):
            self.key, self.value, self.left, self.right = key, value, None, None

File: java/structure/tree/search_tree/BinarySearchTree.java
package structure.tree.search_tree;

public class BinarySearchTree {
    private TreeNode _root = null;

    public boolean isEmpty() {
        return null == _root;
    }

    public void add(int key, int value) {
        if (null == _root) {
            _root = new TreeNode(key, value);
            return;
        }

        TreeNode parent = null;
        TreeNode current = _root;
        while (null != current) {
            if (key == current.key) {
                current.value = value;
                return;
            }

            parent = current;
            if (key < current.key) {
                current = current.left;
            } else {
                current = current.right;
            }
        }
        addNode(parent, key, value);
    }

    public void erase(int key) {
        if (null == _root) {
            return;
        }

        if (key == _root.key && (null == _root.left || null == _root.right)) {
            _root = (null != _root.left) ? _root.left : _root.right;
            return;
        }

        TreeNode parent = null;
        TreeNode current = _root;
        while (null != current) {
            if (key == current.key) {
                break;
            }

            parent = current;
            if (key < current.key) {
                current = current.left;
            } else {
                current = current.right;
            }
        }

        if (null == current) {
            return;
        }

        if (null == current.left || null == current.right) {
            eraseNode(parent, current, (null != current.left) ? current.left : current.right);
            return;
        }

        TreeNode child = current.right;
        parent = current;
        while (null != child.left) {
            parent = child;
            child = child.left;
        }
        current.key = child.key;
        current.value = child.value;
        eraseNode(parent, child, child.right);
    }

    public void set(int key, int value) {
        TreeNode current = _root;
        while (null != current) {
            if (key == current.key) {
                current.value = value;
                return;
            }

            if (key < current.key) {
                current = current.left;
            } else {
                current = current.right;
            }
        }
    }

    public int get(int key) {
        TreeNode current = _root;
        while (null != current) {
            if (key == current.key) {
                return current.value;
            }

            if (key < current.key) {
                current = current.left;
            } else {
                current = current.right;
            }
        }
        return Integer.MIN_VALUE;
    }

    public void clear() {
        _root = null;
    }

    private void addNode(TreeNode parent, int key, int value) {
        if (key < parent.key) {
            parent.left = new TreeNode(key, value);
        } else {
            parent.right = new TreeNode(key, value);
        }
    }

    private void eraseNode(TreeNode parent, TreeNode current, TreeNode child) {
        if (current == parent.left) {
            parent.left = child;
        } else {
            parent.right = child;
        }
    }

    private class TreeNode {
        public int key = 0;
        public int value = 0;
        public TreeNode left = null;
        public TreeNode right = null;

        public TreeNode(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }
}
